### PR TITLE
Add Serial Wire Output (SWO) logger backend

### PR DESCRIPTION
**Pull request by mnkp**
This PR adds Serial Wire Output (SWO) logger backend. SWO is an extension of Serial Wire Debug (SWD) port developed by ARM.

A separate commit adds SWO logger support for SiLabs SoCs. This is required as by default, after reset SWO signal is not connected to GPIO pin on SiLabs SoCs.

Tested on efr32_slwstk6061a evaluation board with J-Link SWO Viewer application from Segger:
```
JLinkSWOViewer -device EFR32FG1PxxxF256 -cpufreq 38400000 -swofreq 875000
```

Also tested on sam_e70_xplained (Atmel SAME70 Xplained) evaluation board.

### CODE REVIEW

**Review by chrta — CHANGES_REQUESTED** [review]
Successfully tested on the EFM32PG12B board. Only one minor typo.

**Comment by chrta** [comment]
You can merge this commit: https://github.com/chrta/zephyr/commit/477874d0330daa366dfe935a5f6ffdeeff82619b into yours to support the EFM32PG12B board.

**Comment by mnkp** [comment]
Thanks @chrta, I've included your commit in this PR. I really don't know how I missed the EFM32PG12B.

**Review by chrta — APPROVED** [review]

File: subsys/logging/Kconfig [hunk 1]
```diff
@@ -302,6 +302,31 @@ config LOG_BACKEND_UART
 	help
 	  When enabled backend is using UART to output logs.
 
+config LOG_BACKEND_SWO
+	bool "Enable Serial Wire Output (SWO) backend"
+	depends on ARM
```
Comment (nordic-krch) [inline subsys/logging/Kconfig]: SWO is optional feature in ARM cortex,

File: subsys/logging/Kconfig [hunk 2]
```diff
@@ -302,6 +302,31 @@ config LOG_BACKEND_UART
...
+	int "Set SWO output frequency"
+	default 0
+	help
+	  Set SWO output frequency. Value 0 will select maximum frequency
+	  supported by the given MCU. Not all debug probes support high
+	  frequency SWO operation. In this case the frequency has to be set
+	  manually.
+
+	  SWO value defined by this option will be configured at boot. Most SWO
+	  viewer programs will configure SWO frequeny when attached to the
```
Comment (nordic-krch) [inline subsys/logging/Kconfig]: frequeny

File: subsys/logging/log_backend_swo.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+#define ITM_PORT_LOGGER       0
+
+/* Set TPIU prescaler for the current debug trace clock frequency. */
+#if CONFIG_LOG_BACKEND_SWO_FREQ_HZ == 0
+#define SWO_FREQ_DIV  1
+#else
+#define SWO_FREQ (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC \
+		  + (CONFIG_LOG_BACKEND_SWO_FREQ_HZ / 2))
+#define SWO_FREQ_DIV  (SWO_FREQ / CONFIG_LOG_BACKEND_SWO_FREQ_HZ)
+#if SWO_FREQ_DIV > 0xFFFF
```
Comment (nordic-krch) [inline subsys/logging/log_backend_swo.c]: it could be changed to `BUILD_ASSERT_MSG` or `#error`. I haven't yet seen such pragma in zephyr.

**Review by nordic-krch — CHANGES_REQUESTED** [review]

File: subsys/logging/log_backend_swo.c (hunk 3 above)
Comment (mnkp) [reply subsys/logging/log_backend_swo.c]: `#pragma message` is used as a missing `#warning` directive. In this case `#error` is indeed more appropriate. I've made the change.

File: subsys/logging/Kconfig (hunk 1 above)
Comment (mnkp) [reply subsys/logging/Kconfig]: I've added HAS_SWO Kconfig option.

File: subsys/logging/Kconfig [hunk 4]
```diff
@@ -302,6 +302,36 @@ config LOG_BACKEND_UART
 	help
 	  When enabled backend is using UART to output logs.
 
+config HAS_SWO
```
Comment (nordic-krch) [inline subsys/logging/Kconfig]: this is SoC feature. I'm not 100% sure where to put it, but maybe here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/Kconfig

Like MPU is also an option for ARM core and it's there.

Comment (mnkp) [reply subsys/logging/Kconfig]: Thanks, I've moved the hidden option as suggested.

**Review by nordic-krch — APPROVED** [review]
lgtm, from logger perspective. I haven't review parts soc specific part.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@mnkp can you please rebase this?

**Pull request merged**